### PR TITLE
1306:Migrate TailSamplingProcessor to new OTLP-based internal data model and add Composite Sampler

### PR DESCRIPTION
**Pull request by vikrambe**
**Description:** Migrate Tail Sampling Processor to the new OTLP-based internal data model and add composite sampler
Adding a feature - Composite Sampler 
**Design Doc - Added support for composite policy in tailsampling processor. This would help in grouping sampling policies and rate limiting them. https://docs.google.com/document/d/10wpIv3TtXgOik05smHm3nYeBX48Bj76TCMxPy8e1NZw/edit#heading=h.ecy5l2puwtp4

**Link to tracking Issue:** 1306

**Testing:** Unit testing

**Documentation:** Refer https://github.com/open-telemetry/opentelemetry-collector/issues/1306, https://docs.google.com/document/d/10wpIv3TtXgOik05smHm3nYeBX48Bj76TCMxPy8e1NZw/edit#heading=h.ecy5l2puwtp4

### CODE REVIEW

**Comment by jpkrohling** [comment]
Aren't we doing changes to this processor only after it's moved to -contrib?

**Comment by vikrambe** [comment]
@jpkrohling Can we merge this PR and move the changes?

**Comment by jpkrohling** [comment]
The OTLP changes are probably required for the move anyway, so, I'd be OK to have them merged before the move. The inclusion of the composite sampler would probably better to have in the new repo. Are you OK in splitting those two changes?

**Comment by vikrambe** [comment]
Sure i can split them.

**Comment by vikrambe** [comment]
@jpkrohling I have removed composite policy changes from this PR. I will open a new PR on contrib repo once this gets moved.

File: processor/samplingprocessor/tailsamplingprocessor/config.go [hunk 1]
```diff
@@ -34,6 +34,8 @@ const (
 	StringAttribute PolicyType = "string_attribute"
 	// RateLimiting allows all traces until the specified limits are satisfied.
 	RateLimiting PolicyType = "rate_limiting"
+	// Composite allows a combination of above policies with rate limiting.
+	Composite PolicyType = "composite"
```
Comment (jpkrohling) [inline processor/samplingprocessor/tailsamplingprocessor/config.go:38]: Shouldn't this be removed as well?

File: processor/samplingprocessor/tailsamplingprocessor/config.go [hunk 2]
```diff
@@ -50,6 +52,12 @@ type PolicyCfg struct {
 	RateLimitingCfg RateLimitingCfg `mapstructure:"rate_limiting"`
 }
 
+// RateAllocation used within composite policy
+type RateAllocationCfg struct {
```
Comment (jpkrohling) [inline processor/samplingprocessor/tailsamplingprocessor/config.go:56]: This should be removed too.

File: processor/samplingprocessor/tailsamplingprocessor/sampling/numeric_tag_filter.go [hunk 3]
```diff
@@ -44,31 +43,53 @@ func NewNumericAttributeFilter(logger *zap.Logger, key string, minValue, maxValu
...
-					return Sampled, nil
+			resource := rs.Resource()
+			if !resource.IsNil() {
+				if v, ok := resource.Attributes().Get(naf.key); ok {
+					value := v.IntVal()
+					if value >= naf.minValue && value <= naf.maxValue {
+						return Sampled, nil
+					}
+				}
+			} //
```
Comment (jpkrohling) [inline processor/samplingprocessor/tailsamplingprocessor/sampling/numeric_tag_filter.go:71]: is this `//` a leftover of something?

File: processor/samplingprocessor/tailsamplingprocessor/sampling/numeric_tag_filter.go [hunk 4]
```diff
@@ -44,31 +43,53 @@ func NewNumericAttributeFilter(logger *zap.Logger, key string, minValue, maxValu
...
-			if v, ok := span.Attributes.AttributeMap[naf.key]; ok {
-				value := v.GetIntValue()
-				if value >= naf.minValue && value <= naf.maxValue {
-					return Sampled, nil
+			resource := rs.Resource()
+			if !resource.IsNil() {
+				if v, ok := resource.Attributes().Get(naf.key); ok {
+					value := v.IntVal()
+					if value >= naf.minValue && value <= naf.maxValue {
+						return Sampled, nil
```
Comment (jpkrohling) [inline processor/samplingprocessor/tailsamplingprocessor/sampling/numeric_tag_filter.go:68]: Isn't this changing the behavior of the filter? Previously, it was acting only on span attributes, not resources, no?

If we really want to change the behavior in this PR, we should then at least extract the filter logic into its own function, receiving a list of attributes. The same logic can then be applied to resource attributes and to span attributes.

File: processor/samplingprocessor/tailsamplingprocessor/processor.go [hunk 5]
```diff
@@ -244,37 +241,50 @@ func (tsp *tailSamplingSpanProcessor) makeDecision(id pdata.TraceID, trace *samp
...
+	return idToSpans
+}
+
+func (tsp *tailSamplingSpanProcessor) processTraces(ctx context.Context, resourceSpans pdata.ResourceSpans, td pdata.Traces) error {
+	// Groupd spans per their traceId to minimize contention on idToTrace
+	idToSpans := tsp.groupSpansByTraceKey(resourceSpans, td)
 
 	var newTraceIDs int64
-	singleTrace := len(idToSpans) == 1
+	// singleTrace := len(idToSpans) == 1
```
Comment (jpkrohling) [inline processor/samplingprocessor/tailsamplingprocessor/processor.go:287]: This can be removed.

**Review by jpkrohling — COMMENTED** [review]
There are still a couple of left overs that don't belong to this PR. Otherwise, it's looking good. I love the fact that you added tests where it was missing!

File: processor/samplingprocessor/tailsamplingprocessor/config.go (hunk 1 above)
Comment (vikrambe) [reply processor/samplingprocessor/tailsamplingprocessor/config.go]: Good catch. Thanks. I have removed it

File: processor/samplingprocessor/tailsamplingprocessor/sampling/numeric_tag_filter.go (hunk 3 above)
Comment (vikrambe) [reply processor/samplingprocessor/tailsamplingprocessor/sampling/numeric_tag_filter.go]: done

File: processor/samplingprocessor/tailsamplingprocessor/processor.go (hunk 5 above)
Comment (vikrambe) [reply processor/samplingprocessor/tailsamplingprocessor/processor.go]: Done

File: processor/samplingprocessor/tailsamplingprocessor/sampling/numeric_tag_filter.go (hunk 4 above)
Comment (vikrambe) [reply processor/samplingprocessor/tailsamplingprocessor/sampling/numeric_tag_filter.go]: Removed : i will create a new PR with the resource attribute filter. For now removed this functionality and its test case from this PR. Thanks

File: processor/samplingprocessor/tailsamplingprocessor/config.go (hunk 2 above)
Comment (vikrambe) [reply processor/samplingprocessor/tailsamplingprocessor/config.go]: Done

**Comment by vikrambe** [comment]
@jpkrohling Thanks for the review. I have taken care of the review comments.

File: processor/samplingprocessor/tailsamplingprocessor/processor.go [hunk 6]
```diff
@@ -244,37 +241,49 @@ func (tsp *tailSamplingSpanProcessor) makeDecision(id pdata.TraceID, trace *samp
...
-			tsp.logger.Warn("Span without valid TraceId", zap.String("SourceFormat", td.SourceFormat))
+func (tsp *tailSamplingSpanProcessor) groupSpansByTraceKey(resourceSpans pdata.ResourceSpans, td pdata.Traces) map[traceKey][]*pdata.Span {
+	idToSpans := make(map[traceKey][]*pdata.Span)
+	ilss := resourceSpans.InstrumentationLibrarySpans()
+	for j := 0; j < ilss.Len(); j++ {
+		ils := ilss.At(j)
+		if ils.IsNil() {
 			continue
 		}
-		tk := traceKey(span.TraceId)
```
Comment (jpkrohling) [inline processor/samplingprocessor/tailsamplingprocessor/processor.go:272]: Is this type (`traceKey`) still needed? Can't we have the `pdata.TraceID` as the key? From what I understand, this is converting the `[]byte` representation of the `TraceID` to a string, only to have it as key to the map.

File: processor/samplingprocessor/tailsamplingprocessor/processor.go [hunk 7]
```diff
@@ -244,37 +241,49 @@ func (tsp *tailSamplingSpanProcessor) makeDecision(id pdata.TraceID, trace *samp
...
-		idToSpans[tk] = append(idToSpans[tk], span)
+		for k := 0; k < ils.Spans().Len(); k++ {
+			span := ils.Spans().At(k)
+			tk := traceKey(span.TraceID().Bytes())
```
Comment (jpkrohling) [inline processor/samplingprocessor/tailsamplingprocessor/processor.go:272]: `HexBytes()` might be more suitable here.

**Review by jpkrohling — APPROVED** [review]
Looks sane to me! Time to get a maintainer to review it now.

cc @tigrannajaryan

File: processor/samplingprocessor/tailsamplingprocessor/processor.go (hunk 6 above)
Comment (vikrambe) [reply processor/samplingprocessor/tailsamplingprocessor/processor.go:272]: correct. let me give it a try...

Comment (vikrambe) [reply processor/samplingprocessor/tailsamplingprocessor/processor.go:272]: runtime error: hash of unhashable type pdata.TraceID : Cant use TraceID as the key.

Comment (jpkrohling) [reply processor/samplingprocessor/tailsamplingprocessor/processor.go:272]: Was worth the attempt :-)

Comment (vikrambe) [reply processor/samplingprocessor/tailsamplingprocessor/processor.go:272]: yup

File: processor/samplingprocessor/tailsamplingprocessor/sampling/time_provider.go [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+// MonotonicClock provides monotonic real clock-based current Unix second.
+// Use it when creating a NewComposite which should measure sample rates
+// against a realtime clock (this is almost always what you want to do,
+// the exception is usually only automated testing where you may want
+// to have fake clocks).
+type MonotonicClock struct{}
+
+func (c MonotonicClock) getCurSecond() int64 {
+	return time.Now().Unix()
+}
```
Comment (tigrannajaryan) [inline processor/samplingprocessor/tailsamplingprocessor/sampling/time_provider.go:34]: Where is this used?

File: processor/samplingprocessor/tailsamplingprocessor/processor.go [hunk 9]
```diff
@@ -244,37 +241,49 @@ func (tsp *tailSamplingSpanProcessor) makeDecision(id pdata.TraceID, trace *samp
...
+	idToSpans := make(map[traceKey][]*pdata.Span)
+	ilss := resourceSpans.InstrumentationLibrarySpans()
+	for j := 0; j < ilss.Len(); j++ {
+		ils := ilss.At(j)
+		if ils.IsNil() {
 			continue
 		}
-		tk := traceKey(span.TraceId)
-		idToSpans[tk] = append(idToSpans[tk], span)
+		for k := 0; k < ils.Spans().Len(); k++ {
```
[the remark this answers is not in the document]
Comment (vikrambe) [reply processor/samplingprocessor/tailsamplingprocessor/processor.go]: agreed. done

File: processor/samplingprocessor/tailsamplingprocessor/processor.go [hunk 10]
```diff
@@ -397,18 +406,13 @@ func (tsp *tailSamplingSpanProcessor) dropTrace(traceID traceKey, deletionTime t
...
-	} else {
-		traceTd = consumerdata.TraceData{
-			Node:     td.Node,
-			Resource: td.Resource,
-			Spans:    spans,
-		}
-	}
+func prepareTraceBatch(rss pdata.ResourceSpans) pdata.Traces {
+	traceTd := pdata.NewTraces()
+	traceTd.ResourceSpans().Resize(traceTd.ResourceSpans().Len() + 1)
```
[the remark this answers is not in the document]
Comment (vikrambe) [reply processor/samplingprocessor/tailsamplingprocessor/processor.go]: correct. i had used this with a traceTd that wasn't empty elsewhere. I have removed traceTd.ResourceSpans().Len() and directly used 1

File: processor/samplingprocessor/tailsamplingprocessor/sampling/time_provider.go (hunk 8 above)
Comment (vikrambe) [reply processor/samplingprocessor/tailsamplingprocessor/sampling/time_provider.go]: This was needed for composite policy which was part of this PR earlier, this is a stale file without composite policy. I have removed it

File: processor/samplingprocessor/tailsamplingprocessor/processor.go [hunk 11]
```diff
@@ -397,18 +406,13 @@ func (tsp *tailSamplingSpanProcessor) dropTrace(traceID traceKey, deletionTime t
...
+	rs := traceTd.ResourceSpans().At(traceTd.ResourceSpans().Len() - 1)
+	rs.Resource().InitEmpty()
+	rss.Resource().CopyTo(rs.Resource())
+	rss.InstrumentationLibrarySpans().CopyTo(rs.InstrumentationLibrarySpans())
```
[the remark this answers is not in the document]
Comment (vikrambe) [reply processor/samplingprocessor/tailsamplingprocessor/processor.go]: thanks, there was a bug fixed it

File: processor/samplingprocessor/tailsamplingprocessor/processor.go [hunk 12]
```diff
@@ -244,37 +241,49 @@ func (tsp *tailSamplingSpanProcessor) makeDecision(id pdata.TraceID, trace *samp
...
-		if err := tsp.processTraces(octd); err != nil {
-			errs = append(errs, err)
+	resourceSpans := td.ResourceSpans()
+	sampledTraceData := pdata.NewTraces()
+	for i := 0; i < resourceSpans.Len(); i++ {
+		resourceSpan := resourceSpans.At(i)
+		if resourceSpan.IsNil() {
+			continue
 		}
+		tsp.processTraces(ctx, resourceSpan, sampledTraceData)
```
[the remark this answers is not in the document]
Comment (vikrambe) [reply processor/samplingprocessor/tailsamplingprocessor/processor.go]: Not really, processTraces never returns an error.it just logs it. Even prior to OTLP migration it was returning nil. I just cleaned it up.

File: processor/samplingprocessor/tailsamplingprocessor/processor.go (hunk 7 above)
Comment (jpkrohling) [reply processor/samplingprocessor/tailsamplingprocessor/processor.go:271]: Yes, sorry, `HexString`. Isn't `traceKey` a string-based type? I thought `HexString` would be safer to use as a string, but I guess it's not a big problem here (unless we log it somewhere).

**Comment by vikrambe** [comment]
@tigrannajaryan Thanks for the review. I have taken care of review comments. Requesting review..

File: processor/samplingprocessor/tailsamplingprocessor/processor.go [hunk 13]
```diff
@@ -244,37 +241,48 @@ func (tsp *tailSamplingSpanProcessor) makeDecision(id pdata.TraceID, trace *samp
...
+				tsp.logger.Warn("Span without valid TraceId")
+			}
+			idToSpans[tk] = append(idToSpans[tk], &span)
+		}
 	}
+	return idToSpans
+}
+
+func (tsp *tailSamplingSpanProcessor) processTraces(resourceSpans pdata.ResourceSpans) error {
+	// Groupd spans per their traceId to minimize contention on idToTrace
```
Comment (jpkrohling) [inline processor/samplingprocessor/tailsamplingprocessor/processor.go:282]: If you are updating the PR for some other reason, you might want to fix this typo: `s/Groupd/Group`

**Review by jpkrohling — APPROVED** [review]

File: processor/samplingprocessor/tailsamplingprocessor/processor.go (hunk 13 above)
Comment (vikrambe) [reply processor/samplingprocessor/tailsamplingprocessor/processor.go]: Done

File: processor/samplingprocessor/tailsamplingprocessor/processor.go [hunk 14]
```diff
@@ -244,37 +241,47 @@ func (tsp *tailSamplingSpanProcessor) makeDecision(id pdata.TraceID, trace *samp
...
 }
 
-func (tsp *tailSamplingSpanProcessor) processTraces(td consumerdata.TraceData) error {
-
-	// Groupd spans per their traceId to minimize contention on idToTrace
-	idToSpans := make(map[traceKey][]*tracepb.Span)
-	for _, span := range td.Spans {
-		if len(span.TraceId) != 16 {
-			tsp.logger.Warn("Span without valid TraceId", zap.String("SourceFormat", td.SourceFormat))
+func (tsp *tailSamplingSpanProcessor) groupSpansByTraceKey(resourceSpans pdata.ResourceSpans) map[traceKey][]*pdata.Span {
```
Comment (jpkrohling) [inline processor/samplingprocessor/tailsamplingprocessor/processor.go:260]: Note: we would probably remove this logic in a next iteration of the processor, especially once https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/1235 is done. For instance, the instrumentation library name isn't being kept by this logic here, and I think it should.

**Review by jpkrohling — APPROVED** [review]
LGTM

File: processor/samplingprocessor/tailsamplingprocessor/processor.go (hunk 14 above)
Comment (vikrambe) [reply processor/samplingprocessor/tailsamplingprocessor/processor.go:260]: Sure

**Comment by vikrambe** [comment]
@jpkrohling Thanks for the review
@tigrannajaryan @dmitryax - can you review and merge this PR.

**Review by tigrannajaryan — APPROVED** [review]

**Pull request merged**